### DISCUSSION TITLE
BackgroundTasks won't run: AttributeError: 'function' object has no attribute 'tasks'

### DISCUSSION THREAD

**Discussion by santosh** [fastapi/fastapi#6362] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import os
import time

from dotenv import load_dotenv
from fastapi import FastAPI, Request
from fastapi.background import BackgroundTasks
from fastapi.middleware.cors import CORSMiddleware
from redis_om import get_redis_connection, HashModel
import requests

load_dotenv()

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3000"],
    allow_methods=["*"],
    allow_headers=["*"],
)


redis = get_redis_connection(
    host="redis-14242.c212.ap-south-1-1.ec2.cloud.redislabs.com",
    port=14242,
    password=os.getenv("REDIS_LABS_PASSWORD", ""),
    decode_responses=True,
)

class Order(HashModel):
    product_id: str
    price: float
    fee: float
    total: float
    quantity: int
    status: str

    class Meta:
        database = redis

@app.get("/orders/{pk}")
def get(pk: str):
    return Order.get(pk)

@app.post("/orders")
async def create(request: Request, background_tasks=BackgroundTasks):
    body = await request.json()
    # req = requests.get("http://localhost:7777/products/%s" % body["id"])
    
    # product = req.json()

    product = {
        "id": "01FZVV98SJ90R5WN11P9Y2RXR5",
        "name": "product",
        "price": 20.0,
        "quantity": 100
    }

    order = Order(
        product_id = body["id"],
        price = product["price"],
        fee = 0.2 * product["price"],
        total = 0.2 * product["price"] + product["price"],
        quantity=body["quantity"],
        status="pending"
    )

    order.save()

    background_tasks.add_task(order_completed, order)

    return order


def order_completed(order: Order):
    print("processing payment")
    time.sleep(5)
    print("done processing payment")
    order.status = 'completed'
    order.save()
```

### Description

* Prerequisite: Create a free instance of redis db from redislabs (or run locally, whichever is easier). Check the requirements.txt I have provided later in this post.

1. There are 2 endpoints: `POST /orders` to create an order, and `GET /orders/{pk}` to check the status of an order.
2. In `create` function, which is linked to `POST /orders`, we initially set order status to `pending`. 
3. We have an independent function that takes in the order and set the order status to `completed`. This function sleeps for 5 secs before doing that transition.

4. You have to send a POST request with this similar body:
```json
{
    "id": "01FZVV98SJ90R5WN11P9Y2RXR5",
    "quantity": 2
}
```

When the execution flow reaches `background_tasks.add_task(order_completed, order)`, it throws an exception:

```
  File "/home/santosh/workspace/faststore/payment/.venv/lib/python3.10/site-packages/starlette/background.py", line 31, in add_task
    self.tasks.append(task)
AttributeError: 'function' object has no attribute 'tasks'
```

---
Additional info:

I read through declaration of `BackgroundTasks` and found that `self.tasks` is only created when `__init__` is called. So I changed this:

```async def create(request: Request, background_tasks=BackgroundTasks):```
to this:

```async def create(request: Request, background_tasks=BackgroundTasks()):````

(not that I instanciated the `BackgroundTasks`). 

This no longer throws the exception, but the background task of changing the order status is never run and the order stays in pending state. 

### Operating System

Linux

### Operating System Details

Linux kubuntu 5.13.0-39-generic #44~20.04.1-Ubuntu SMP Thu Mar 24 16:43:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### FastAPI Version

0.75.1

### Python Version

Python 3.10.2

### Additional Context

Output of pip freeze:

```
aioredis==2.0.1
anyio==3.5.0
asgiref==3.5.0
async-timeout==4.0.2
certifi==2021.10.8
charset-normalizer==2.0.12
cleo==1.0.0a4
click==8.1.2
crashtest==0.3.1
Deprecated==1.2.13
fastapi==0.75.1
h11==0.13.0
hiredis==2.0.0
idna==3.3
packaging==21.3
pptree==3.1
pydantic==1.9.0
pylev==1.4.0
pyparsing==3.0.7
python-dotenv==0.20.0
python-ulid==1.1.0
redis==4.2.2
redis-om==0.0.21
requests==2.27.1
six==1.16.0
sniffio==1.2.0
starlette==0.17.1
types-redis==4.1.19
types-six==1.16.12
typing_extensions==4.1.1
urllib3==1.26.9
uvicorn==0.17.6
wrapt==1.14.0
```

**Comment by tasercake** [comment, marked as answer]
### One-line Fix

In your `/orders` view function, use `background_tasks: BackgroundTasks` instead of `background_tasks=BackgroundTasks`

```python
@app.post("/orders")
async def create(request: Request, background_tasks: BackgroundTasks):
	...
```

---

### Why your code didn't work

[`BackgroundTasks`](https://github.com/encode/starlette/blob/20d24a86b877e21ed002abb1668e69002850f38b/starlette/background.py#L31-L43) is a class that should be instantiated before use, so `background_tasks=BackgroundTasks` won't do quite what you expect it to.

FastAPI uses [Dependency Injection](https://fastapi.tiangolo.com/tutorial/dependencies/) to provide you with an _instance_ of `BackgroundTasks` ([source](https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/dependencies/utils.py#L566-L569)) that's [automatically queued](https://github.com/encode/starlette/blob/20d24a86b877e21ed002abb1668e69002850f38b/starlette/responses.py#L160-L163) by Starlette after your view function returns a response.

(I'd encourage you to read through the linked source code to firm up your understanding of the problem)

If you look at Starlette's implementation of `add_task` you'll see something like:

```python
class BackgroundTasks(...):
	...
    def add_task(self, func, *args, **kwargs):
        task = BackgroundTask(func, *args, **kwargs)
        self.tasks.append(task)
```

Where `self` is meant to be the instance of `BackgroundTasks` injected by FastAPI.

In your current code (where `background_tasks=BackgroundTasks`) when you call `background_tasks.add_task(order_completed, order)` you're effectively doing:

```python
BackgroundTasks.add_task(order_completed, order)
# Which effectively does:
order_completed.tasks.append(BackgroundTask(order))
```

In the 'fixed' code (where `background_tasks` is an _instance_ of `BackgroundTasks`) when you do `background_tasks.add_task(order_completed, order)` you'd effectively be doing:

```python
BackgroundTasks.add_task(background_tasks, order_completed, order)
# Which effectively does:
background_tasks.tasks.append(BackgroundTask(order_completed, order))
```

See the FastAPI docs on [Background Tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/) for more info.

**Comment by santosh** [comment]
My bad. I'm not much used to type annotations. Will pay extra attention to it.
